Excluding OS Specific Dependencies Fails with ModuleNotFoundError

In https://github.com/gaphor/gaphor/pull/2045, I tried adding a dependency just for macOS:
`pyobjc = {version = "^9.0.1", markers = "sys_platform == 'darwin'"}`

We are then making use of the import in our i18n module:

```
    if sys.platform == "darwin":
        from Cocoa import NSUserDefaults

        defaults = NSUserDefaults.standardUserDefaults()
        langs = defaults.objectForKey_("AppleLanguages")
        if language := langs.objectAtIndex_(0):
            return language
```

Then we are testing it with pytest-archon like:

```
def test_core_packages():
    (
        archrule("gaphor.core does not depend on the rest of the system")
        .match("gaphor.core*")
        .exclude("*.tests.*")
        .may_import(*GAPHOR_CORE)
        .should_not_import("gaphor*")
        .should_not_import(*UI_LIBRARIES)
        .should_not_import("Cocoa")
        .check(gaphor, skip_type_checking=True)
    )
```

However this results in:

```
    def test_core_packages():
        (
            archrule("gaphor.core does not depend on the rest of the system")
            .match("gaphor.core*")
            .exclude("*.tests.*")
            .may_import(*GAPHOR_CORE)
            .should_not_import("gaphor*")
            .should_not_import(*UI_LIBRARIES)
            .should_not_import("Cocoa")
>           .check(gaphor, skip_type_checking=True)
        )

tests\test_architecture.py:37:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  
.venv\Lib\site-packages\pytest_archon\rule.py:176: in check
    all_imports = collect_imports(
.venv\Lib\site-packages\pytest_archon\collect.py:35: in collect_imports
    for name, imports in collect_imports_from_path(package_dir(package), package, walker):
.venv\Lib\site-packages\pytest_archon\collect.py:85: in collect_imports_from_path
    return frozenset(_collect(py_file) for py_file in Path(path).glob("**/*.py"))
.venv\Lib\site-packages\pytest_archon\collect.py:85: in <genexpr>
    return frozenset(_collect(py_file) for py_file in Path(path).glob("**/*.py"))
.venv\Lib\site-packages\pytest_archon\collect.py:83: in _collect
    return module_name, frozenset(imports)
.venv\Lib\site-packages\pytest_archon\collect.py:113: in extract_imports_ast
    yield resolve_module_or_object_by_path(fqname)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

fqname = 'Cocoa.NSUserDefaults'

    def resolve_module_or_object_by_path(fqname: str) -> str:
        if "." not in fqname:
            return fqname

        if fqname in core_modules():
            return fqname

        parts = fqname.split(".")
        head = parts[0]
        parent_name = ".".join(parts[:-1])

        # shortcut to deal with e.g. from __future__ import annotations
        if parent_name in core_modules():
            return parent_name

        # taken from importlib.util.find_spec
        if fqname in sys.modules:
            module = sys.modules[fqname]
            if module is None:
                return None
            try:
                spec = module.__spec__
            except AttributeError:
                raise ValueError(f"{fqname}.__spec__ is not set") from None
            else:
                if spec is None:
                    raise ValueError(f"{fqname}.__spec__ is None")
                return fqname

        spec = find_spec(head)
        if not spec:
>           raise ModuleNotFoundError(f"could not find the module {head} to resolve {fqname}", name=head)
E           ModuleNotFoundError: could not find the module Cocoa to resolve Cocoa.NSUserDefaults

.venv\Lib\site-packages\pytest_archon\collect.py:212: ModuleNotFoundError
```

Since it isn't possible to install pyobjc in Windows or Linux, it would be nice to be able to either exclude the module entirely or only try to find the module if it passes the `sys.platform` check.

--- a/pytest_archon/collect.py
+++ b/pytest_archon/collect.py
@@ -10,6 +10,7 @@ from importlib.util import find_spec
 from pathlib import Path
 from types import ModuleType
 from typing import Callable, Dict, Iterator, Set, Iterable, Sequence
+from logging import getLogger
 
 from pytest_archon.core_modules import core_modules
 
@@ -20,9 +21,9 @@ from pytest_archon.core_modules import core_modules
 
 
 Walker = Callable[[ast.Module], Iterator[ast.AST]]
-
-
 ImportMap = Dict[str, Set[str]]
+
+logger = getLogger(__name__)
 
 
 def collect_imports(package: str | ModuleType, walker: Walker) -> ImportMap:
@@ -113,6 +114,10 @@ def extract_imports_ast(nodes: Iterator[ast.AST], package: str, resolve=True) ->
                     try:
                         yield resolve_module_or_object_by_path(fqname)
                     except ModuleNotFoundError:
+                        logger.warning(
+                            f"Could not determine if {fqname} is a module or function."
+                            " Assuming it is a module."
+                        )
                         yield fqname
                 else:
                     yield fqname

--- a/pytest_archon/collect.py
+++ b/pytest_archon/collect.py
@@ -110,7 +110,10 @@ def extract_imports_ast(nodes: Iterator[ast.AST], package: str, resolve=True) ->
             for alias in node.names:
                 fqname = resolve_import_from(alias.name, node.module, package=package, level=node.level)
                 if resolve:
-                    yield resolve_module_or_object_by_path(fqname)
+                    try:
+                        yield resolve_module_or_object_by_path(fqname)
+                    except ModuleNotFoundError:
+                        yield fqname
                 else:
                     yield fqname
 

--- a/tests/test_rule.py
+++ b/tests/test_rule.py
@@ -192,3 +192,13 @@ def test_forbidden_transitive_dependency_fails(create_testset):
     assert "- module 'abcz.moduleA' has FORBIDDEN import" in longrepr
     assert "abcz.moduleD (matched by glob pattern /abcz.moduleD/)" in longrepr
     assert "abcz.moduleA ↣ abcz.moduleB ↣ abcz.moduleC ↣ abcz.moduleD" in longrepr
+
+
+def test_resolution_non_existing_module(create_testset):
+    create_testset(
+        ("abcz/__init__.py", ""),
+        ("abcz/moduleA.py", "import abcz.moduleB"),
+        ("abcz/moduleB.py", "from i_do_not_exist import i_also_do_not_exist\nimport abcz.moduleC"),
+        ("abcz/moduleC.py", ""),
+    )
+    archrule("rule exclusion").match("abcz.moduleA").should_import("abcz.moduleC").check("abcz")

